### ISSUE TITLE
make test not pass when enable KTLS

### ISSUE THREAD

**Issue by chuanshan-ali** [openssl/openssl#9648]
I am using the latest features of openssl. I found that recently openssl has started to support KTLS. So here my steps:
1. download latest stable kernel and turn on kernel TLS
2. compile and install the kernel
3. `cd openssl` `./config enable-ktls` `make -j6 && make test`

But `sslapitest` cannot pass. Here is error:
```
../test/recipes/90-test_sslapi.t                 (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
```

**Comment by chuanshan-ali** [comment]
BTW.

- KTLS is merged upstream when 4.13, So I also tried directly `sudo modprobe tls` on ubuntu server 18.04.
- just `./config` and `make test` pass all tests

**Comment by chuanshan-ali** [comment]
minimal kernel version：4.17

**Comment by chuanshan-ali** [comment]
update kernel cannot solve this problem
![image](https://user-images.githubusercontent.com/54096794/63664564-9c9f9900-c7f9-11e9-8dda-3cacabfdda1d.png)

**Comment by mattcaswell** [comment]
@borisPis

**Comment by mattcaswell** [comment]
@borisPis

**Comment by dcbeckett** [comment]
Not sure if this is the cause of your issue, but there is a open bug relating to ktls

https://github.com/openssl/openssl/issues/8896

To avoid hitting this bug with my test environment, I usually compile openssl off of the commit before the bug was introduced (git checkout 8450d0)

**Comment by jayaram-intel** [comment]
Hi @chuanshan-ali ,

I am trying to reproduce the above issue with the Linux kernel 5.2.11.
Please provide steps in detail about how you enabled Kernel TLS support, So that I will try with same options on my side.

**Comment by jayaram-intel** [comment]
Hi @chuanshan-ali ,
I tried to reproduce the above issue with the Linux kernel 5.2.11.
As of now I performed below steps:
1.	Downloaded Linux kernel 5.2.11 Enabled kernel TLS ( CONFIG_TLS=m and CONFIG_TLS_DEVICE=y)
2.	Compiled and installed the 5.2.11 kernel
3.	After cloning latest Openssl code base performed below steps
A ./config '-WL,--enable-new-dtags,-rpath,$(LIBRPATH)' enable-ktls
B. make
C. make test
            **I didn’t seen any test case fail issue, sslapi test is also passed.**
             ../test/recipes/90-test_sslapi.t …………..ok
D. sudo make install

Please let me know in case if I missed any steps or configurations need  to enable.

**Comment by chuanshan-ali** [comment]
Hi @jayaram-intel ,
I noticed that your CONFIG_TLS is optional. Did you check `lsmod | grep tls` and verify that KTLS is enabled?
In my steps, just `sudo modprobe tls` and test. Because this feature has merged into mainline after 4.13 and added `TLS_RX` after 4.17. My kernel is 5.0 (in ubuntu 18.04 and `sudo apt install linux-image-5.0.0-27-generic`)

**Comment by jayaram-intel** [comment]
Hi @chuanshan-ali ,

1. Inserted tls module using `sudo modprobe tls`
2. Verified KTLS using `lsmod | grep tls` , yes it is enabled
3. After that compiled Openssl using below commands
     -  ./config '-WL,--enable-new-dtags,-rpath,$(LIBRPATH)' enable-ktls
     - make
     - make TESTS="test_sslapi" test VERBOSE=1
Observed that **around 17 cases are not ok**
Please find below attached log file : test_sslapi_maketest_fail_log.txt
[test_sslapi_maketest_fail_log.txt](https://github.com/openssl/openssl/files/3599184/test_sslapi_maketest_fail_log.txt)

**Comment by jayaram-intel** [comment]
Hi @chuanshan-ali ,

I have gone through KTLS related commits available in both **OpenSSL Master branch** and **Linux Kernel** (till 16th September 2019).

**Identified that  KTLS related breaks are there after Linux kernel version 4.20.**

**I am able to Compile and Make test is passed with Linux kernel version 4.20.**

**Please try to use Linux kernel version 4.20 and verify from your end also.**

**This issue is purely related to Linux kernel side KTLS breaks, requesting you to close this issue.**

**Note:**
_Make test is failed in Linux kernel versions 5.0 and 5.2.11 due to the Linux kernel side KTLS breaks._

**Comment by hunterlxt** [comment]
@jayaram-intel 
I have not been internship there. But I will go to verify according to your method.

**Comment by jayaram-intel** [comment]
Hi @hunterlxt ,
Sure, you can verify.
BTW,
Just now I cross verified with **different Linux kernel versions.**
Able to compile and **`make test`** is passed for **`Linux kernel versions 4.20 and 5.2.`**
And **`make test`** is failed for **`Linux kernel versions 5.0 and 5.2.11.`**

Note:
Linux Kernel TLS is enabled using `sudo modprobe tls`

Thanks
Jayaram Matta

**Comment by BorisPis** [comment]
Hi all,

I could not reproduce the kernel issue, but I do see an openssl issue related to [PR8793](https://github.com/openssl/openssl/pull/8793) which modified BIO_get_ktls_send and BIO_get_ktls_recv to verify the BIO_method_type to be BIO_TYPE_SOCKET.

This breaks KTLS support as KTLS is often set on a FD or any other method of providing a socket OpenSSL.

After reverting this change the tests on master pass with the stable net tree: git://git.kernel.org/pub/scm/linux/kernel/git/davem/net.git

**Comment by jayaram-intel** [comment]
Hi @BorisPis,

I need your response on below items

> I could not reproduce the kernel issue, but I do see an openssl issue related to [PR8793](https://github.com/openssl/openssl/pull/8793)

On which kernel versions you tried it ( whether you tried it only on stable net tree: git://git.kernel.org/pub/scm/linux/kernel/git/davem/net.git only after reverting [PR8793 ](https://github.com/openssl/openssl/pull/8793)) or you tried with any other Linux kernel versions like 5.0 and 5.2.11.

> After reverting this change the tests on master pass with the stable net tree: git://git.kernel.org/pub/scm/linux/kernel/git/davem/net.git

From your above comment I understood that, If we revert the [PR8793](https://github.com/openssl/openssl/pull/8793) changes then we will not see any OpenSSL `make test` fail issues on any latest kernel versions (latest stable and also 5.0 and 5.2.11 versions) right?. Correct me if my understanding is wrong.

In case if my understanding is right, still I am facing OpenSSL `make test` fail issue with Linux kernel version 5.2.11 (tried with/without Reverting [PR8793](https://github.com/openssl/openssl/pull/8793)).




Thanks and Regards,
Jayaram Matta

**Comment by BorisPis** [comment]
> From your above comment I understood that, If we revert the PR8793 changes then we will not see any OpenSSL make test fail issues on any latest kernel versions (latest stable and also 5.0 and 5.2.11 versions) right?. Correct me if my understanding is wrong.

I haven't tried reverting, but I removed the lines mentioned above and posted a fix in [PR9973](https://github.com/openssl/openssl/pull/9973)

Can you try with the stable net tree? It is possible there are some fixes that haven't been backported to 5.2.11.

KTLS development is still active, and you can run the tests in the kernel to verify that it works: https://github.com/torvalds/linux/blob/master/tools/testing/selftests/net/tls.c

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input
